### PR TITLE
Private recipes

### PR DESCRIPTION
**Pull request by dougnd**
I've started hosting this locally for my family -- it is quite nice!  I got a request for the ability to have private recipes (only logged in users could see them).   This is my first attempt and I am new to Django REST and some of the technologies here so it may need work before you accept.   

The privacy may not be completely secure (i.e. you can still view the pictures with the URL), but it works for my use case.

I am open to any suggestions!

### CODE REVIEW

**Comment by RyanNoelk** [comment]
Thanks for the PR!

I took a quick look and for the most part things looks good. I have some more detailed comments later. but for now I noted you are missing one big thing.

You need to make sure that for each filter, you are reducing the queryset appropriately. Meaning that when a users does a search on the browse page, the recipe count for each filter is displayed properly (depending on whether or not they are logged in).

- https://github.com/RyanNoelk/OpenEats/blob/dev/api/v1/recipe_groups/views.py#L42
- https://github.com/RyanNoelk/OpenEats/blob/dev/api/v1/recipe_groups/views.py#L95
- https://github.com/RyanNoelk/OpenEats/blob/dev/api/v1/recipe/views.py#L92

**Comment by dougnd** [comment]
Good point!  Those counts should now be fixed.

I was also wondering how it would affect sub-recipes.  But I am not sure I understand that feature. The "Recipe links" section is either cleared/ignored on save unless a line is a previous recipe name at which point the recipe POST/PATCH api call dies and returns a 500 error.  Is that still WIP?   I can post details if you like.

**Comment by RyanNoelk** [comment]
Sub-recipes should be un-effected by this. The only way you would see a private recipe is if a user added a private recipe as a sub-recipe to a public recipe.

As for sub-recipes working. They are an experimental feature right now. If you have have issues with it, please open a ticket and we can discuss if more there.

Another random thought I had, should a user still be able to access the recipe directly or should that be blocked? Meaning if I just type in a url with a random recipe ID, should a user be able to access it if it's private. We can always deal with this case later if an issue arises.

File: api/v1/recipe/views.py [hunk 1]
```diff
@@ -92,6 +104,10 @@ class RatingViewSet(viewsets.ReadOnlyModelViewSet):
     def get_queryset(self):
         query = Recipe.objects
 
+        # If user is anonymous, restrict recipes to public.
+        if not self.request.user.is_authenticated:
+            query = query.filter(public=True)
```
Comment (RyanNoelk) [inline api/v1/recipe/views.py]: This should be added to the `filter_set` below. 

Something like this should work:
``` python
filter_set = {}
if not self.request.user.is_authenticated:
    filter_set['public'] = True
```

Chaining together filters can be weird in django.

File: api/v1/recipe_groups/views.py [hunk 2]
```diff
@@ -95,6 +99,10 @@ class CourseCountViewSet(viewsets.ModelViewSet):
     def get_queryset(self):
         query = Recipe.objects
 
+        # If user is anonymous, restrict recipes to public.
+        if not self.request.user.is_authenticated:
+            query = query.filter(public=True)
+
```
Comment (RyanNoelk) [inline api/v1/recipe_groups/views.py]: Same comment as above.

Something like this should work:
``` python
filter_set = {}
if not self.request.user.is_authenticated:
    filter_set['public'] = True
```

File: api/v1/recipe_groups/views.py [hunk 3]
```diff
@@ -42,6 +42,10 @@ class CuisineCountViewSet(viewsets.ModelViewSet):
...
```
Comment (RyanNoelk) [inline api/v1/recipe_groups/views.py]: Same comment as above.

Something like this should work:
``` python
filter_set = {}
if not self.request.user.is_authenticated:
    filter_set['public'] = True
```

**Review by RyanNoelk — CHANGES_REQUESTED** [review]

**Comment by dougnd** [comment]
Good to know about chaining filters.  Should be fixed now.

As for whether a user should be able to access the recipe if they know the URL, I figured I'd deny that too.  I could see ideally there be more options:  Public, Private, Hidden, and even user groups where a recipe can be restricted to the group.  But for my purposes, Private and Public are all I need.

**Review by RyanNoelk — APPROVED** [review]
LGTM!

**Pull request merged**